### PR TITLE
Generate LigCaretList table

### LINKED ISSUE #809: Generate ligature caret table
This is something that fontmake does for oswald but that we don't. Closely related to #805, in pythonland both of these things are handled by the [`GdefFeatureWriter`](https://github.com/googlefonts/ufo2ft/blob/main/Lib/ufo2ft/featureWriters/gdefFeatureWriter.py)

### PR DESCRIPTION
**Pull request by cmyr**
~*This is currently patching fontations to bring in https://github.com/googlefonts/fontations/pull/1090, I'll do a release when that gets merged*~

This patch generates a `LigCaretList` entry for any glyphs which include an anchor name beginning with `caret_` or `vcaret_`.

- currently this happens during mark generation, which is convenient since we have access to all the anchors there.
- this matches the `LigCaretList` produced by `fontmake` for oswald.
- closes #809

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -53,3 +53,7 @@ members = [
     "otl-normalizer",
     "fontc_crater",
 ]
+
+[patch.crates-io]
```
Comment (rsheeter) [inline Cargo.toml:57]: I think we should view this as submit-blocking, lets just release fontations.

File: fea-rs/src/compile/compile_ctx.rs [hunk 2]
```diff
@@ -331,6 +333,10 @@ impl<'a, F: FeatureProvider, V: VariationInfo> CompilationCtx<'a, F, V> {
             gdef.mark_glyph_sets = sorted.into_iter().map(|(_, cls)| cls).collect();
         }
 
+        if gdef.ligature_pos.is_empty() {
+            gdef.ligature_pos = user_lig_carets;
+        }
```
Comment (rsheeter) [inline fea-rs/src/compile/compile_ctx.rs:338]: else ... warn if the user had specified carets? Or comment to explain why just ignoring the user carets is fine.

File: Cargo.toml (hunk 1 above)
Comment (cmyr) [reply Cargo.toml:57]: totally, just waiting for a review.

File: fea-rs/src/compile/compile_ctx.rs (hunk 2 above)
Comment (cmyr) [reply fea-rs/src/compile/compile_ctx.rs:338]: oops this was just a bad name, the `user_lig_carets` are the ones generated by the compiler.

File: fea-rs/src/compile/tables/gdef.rs [hunk 3]
```diff
@@ -40,17 +43,26 @@ pub struct GdefBuilder {
...
+            self.build_lig_caret_list(&mut var_store),
             self.build_mark_attach_class_def(),
         );
 
         table.mark_glyph_sets_def = self.build_mark_glyph_sets().into();
-        if let Some((var_store, key_map)) = self.var_store.clone().map(VariationStoreBuilder::build)
-        {
+        // only use the var_store if we had one set at the start of this fn
+        let var_store = Some(var_store).filter(|_| self.var_store.is_some());
+        if let Some((var_store, key_map)) = var_store.map(VariationStoreBuilder::build) {
```
Comment (rsheeter) [inline fea-rs/src/compile/tables/gdef.rs:63]: Creating a mutable varstore that isn't self.var_store and then having to carefully remember if it's bogus or not seems fairly ugly. Why is that necessary? We can't or shouldn't code it to run against self.var_store and keep the original code here?

File: fontbe/src/features/marks.rs [hunk 4]
```diff
@@ -621,6 +627,62 @@ fn resolve_anchor_once(
     Ok(anchor)
 }
 
+fn get_ligature_carets(
```
Comment (rsheeter) [inline fontbe/src/features/marks.rs:630]: Is there a python equivalent to link to?

I suggest we should have a comment somewhere, perhaps here, that captures, or links to, the discussion of whether we need the table, why / who uses it.

File: fea-rs/src/compile/tables/gdef.rs (hunk 3 above)
Comment (rsheeter) [reply fea-rs/src/compile/tables/gdef.rs:63]: At a glance it seems like it might be ok for build to take `&mut self`, as CompilationCtx::build - which appears to be the primary caller - does?

Comment (cmyr) [reply fea-rs/src/compile/tables/gdef.rs:63]: agree that this is annoying, and is the consequence of a few things colliding:
- we need to mutate the var store to add deltas to it, in `build_lig_caret_list`.
- for Annoying Rust Reasons, `Option<&mut T>` is not `Copy`. This means you can't do something like,

```rust
let mut thing = Thing;
let thing_ref = Some(&mut Thing);
for _ in 0..5 {
    do_thing(thing_ref); // < error: `thing_ref` moved in previous loop pass
}
```
- and as a consequence of this, `build_lig_caret_list` (along with many other similar functions) take `&mut VariationStoreBuilder` instead of `Option<&mut VariationStoreBuilder>`
- and this means that even if there is no varstore at this point (and this is late in compilation, if this was a variable font we would have a varstore) we need to create one in order to call that method; but if we created one just for this we don't care to keep it on, and so we discard it afterwards.

having `&mut self` wouldn't really change anything; we would still need to create a dummy varstore in order to build the lig caret list, and we would still discard it afterwards.

Comment (rsheeter) [reply fea-rs/src/compile/tables/gdef.rs]: Per IM checking if the varstore is empty instead of `Some(var_store).filter(|_| self.var_store.is_some())` should suffice.

**Review by rsheeter — APPROVED** [review]
LGTM subject to `Some(var_store).filter(|_| self.var_store.is_some())` switching to just a check for var_store not being empty. One hopes 0 axis varstores are always empty since no delta that makes sense seems possible.

**Comment by cmyr** [comment]
changed to a simpler version of this original check; if for some crazy reason there were no variations for any values in gpos but there happen to be variations in the ligature carets the empty check would fail, and we'd compile garbage, so I prefer to just always chuck the var_store if we have no deltas at this point.
